lib/open_telemetry/honeycomb/json_backend.ex

```diff
@@ -0,0 +1,73 @@
+defmodule OpenTelemetry.Honeycomb.JsonBackend do
+  @moduledoc """
+  JSON back end.
+
+  The OpenTelemetry Honeycomb Exporter uses a Jason/Poison-style JSON encoder via a behaviour so
+  you can adapt it to your preferred JSON encoder or decode/encode options.
+  """
+
+  alias OpenTelemetry.Honeycomb.JsonBackend.PoisonBackend
+
+  @typedoc "A term that we can encode to JSON, or decode from JSON."
+  @type encodable_term ::
+          nil
+          | boolean()
+          | float()
+          | integer()
+          | String.t()
+          | [encodable_term]
+          | %{optional(String.t()) => encodable_term()}
+
+  @doc """
+  Decode a term from JSON.
+  """
+  @callback decode!(iodata()) :: encodable_term() | no_return()
+
+  @doc """
+  Encode a term to JSON.
+  """
+  @callback encode_to_iodata!(encodable_term()) :: iodata() | no_return()
+
+  @doc """
+  Return the default back end's module.
+  """
+  def default_module, do: PoisonBackend
+
+  @typedoc "Configuration for `decode!/2` and `encode_to_iodata!/2`."
+  @type config :: %{
+          required(:json_module) => module(),
+          optional(any()) => any()
+        }
```

Comment:
You're not using any other keys in this map. Does it derive somewhere from the application-level configuration? Your `Config` module?

Reply:
I had to break the circular relationship for some reason or other, and Dialyzer insisted on the `_ => _`. So annoying. Composing the option typings out of keyword/value pairs without the wrapped list and then composing the main options as a combination of these would do the trick.

Reply:
Not sure I like how it turned out. It'd be tidier if ExDoc hadn't insisted on the unaliased module name in the type reference, but even then we'd still be telling people to look in three places instead of one:

```elixir
  @typedoc """
  Configuration for the OpenTelemetry Honeycomb exporter, giving:

  * `api_endpoint`: the API endpoint
  * `attribute_map`: a map to control dataset attributes used for span properties (see below)
  * `dataset`: the Honeycomb dataset name
  * `write_key`: the write key

  Please see also the options controlling how the exporter encodes JSON and makes HTTP requests:

  * `t:OpenTelemetry.Honeycomb.HttpBackend.config_opt/0`
  * `t:OpenTelemetry.Honeycomb.JsonBackend.config_opt/0`

  If the `write_key` is absent or `nil`, the exporter won't attempt any outbound requests.
  """
  @type config_opt ::
          {:api_endpoint, String.t()}
          | {:attribute_map, AttributeMap.t()}
          | {:dataset, String.t()}
          | {:write_key, String.t() | nil}
          | HttpBackend.config_opt()
          | JsonBackend.config_opt()
```

Reply:
I felt clever for a split second, here:
```
  #{
    OpenTelemetry.Honeycomb.HttpBackend
    |> Code.fetch_docs()
    |> elem(6)
    |> Enum.find_value(fn
      {{:type, :config_opt, 0}, _, _, %{"en" => doc}, _} -> doc
      _ -> nil
    end)
  }
```
… but compilation order. Dang.

Reply:
Nuts. Forgot the struct also gives me a ripper way of defining defaults. Bulking up on that, too, to see how it turns out…

Reply:
Pushed changes to make the main config a keyword list. I like the feel of pulling the types and defaults for the back ends from the back ends, but the rest feels ugly. Maybe more square brackets and less `Keyword.fetch!/2`…